File: app/src/main/java/com/example/bloodbank/MyAppointmentsActivity.java
package com.example.bloodbank;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.bloodbank.adapters.AppointmentsAdapter;
import com.example.bloodbank.models.Donation;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class MyAppointmentsActivity extends AppCompatActivity {
    private RecyclerView appointmentsRecyclerView;
    private AppointmentsAdapter adapter;
    private List<Donation> appointmentsList;
    private SwipeRefreshLayout swipeRefreshLayout;
    private View progressBar;
    private TextView emptyView;
    private DatabaseReference donationsRef;
    private String userId;
    private ValueEventListener appointmentsListener;
    private Query appointmentsQuery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_my_appointments);

        // Initialize Firebase
        userId = FirebaseAuth.getInstance().getCurrentUser().getUid();
        donationsRef = FirebaseDatabase.getInstance().getReference().child("donations");

        // Initialize views
        initializeViews();
        
        // Setup toolbar
        setupToolbar();
        
        // Setup RecyclerView
        appointmentsList = new ArrayList<>();
        adapter = new AppointmentsAdapter(appointmentsList);
        appointmentsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        appointmentsRecyclerView.setAdapter(adapter);
        
        // Setup SwipeRefreshLayout
        swipeRefreshLayout.setOnRefreshListener(this::loadAppointments);
        
        // Load appointments
        loadAppointments();
    }
    
    private void initializeViews() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        appointmentsRecyclerView = findViewById(R.id.appointmentsRecyclerView);
        swipeRefreshLayout = findViewById(R.id.swipeRefreshLayout);
        progressBar = findViewById(R.id.progressBar);
        emptyView = findViewById(R.id.emptyView);
    }
    
    private void setupToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle("My Appointments");
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
        }
    }
    
    private void loadAppointments() {
        progressBar.setVisibility(View.VISIBLE);
        emptyView.setVisibility(View.GONE);
        
        // Remove previous listener if exists
        if (appointmentsListener != null && appointmentsQuery != null) {
            appointmentsQuery.removeEventListener(appointmentsListener);
        }
        
        // Query donations for the current user
        appointmentsQuery = donationsRef.orderByChild("donorId").equalTo(userId);
        
        appointmentsListener = new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                appointmentsList.clear();
                
                if (dataSnapshot.exists()) {
                    for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                        Donation donation = snapshot.getValue(Donation.class);
                        if (donation != null) {
                            donation.setDonationId(snapshot.getKey());
                            appointmentsList.add(donation);
                        }
                    }
                }
                
                // Sort appointments by date (newest first)
                appointmentsList.sort((d1, d2) -> Long.compare(d2.getTimestamp(), d1.getTimestamp()));
                
                // Update UI
                adapter.notifyDataSetChanged();
                progressBar.setVisibility(View.GONE);
                swipeRefreshLayout.setRefreshing(false);
                
                // Show empty view if no appointments
                if (appointmentsList.isEmpty()) {
                    emptyView.setVisibility(View.VISIBLE);
                } else {
                    emptyView.setVisibility(View.GONE);
                }
            }
            
            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                progressBar.setVisibility(View.GONE);
                swipeRefreshLayout.setRefreshing(false);
                Toast.makeText(MyAppointmentsActivity.this, 
                        "Error loading appointments: " + databaseError.getMessage(), 
                        Toast.LENGTH_SHORT).show();
            }
        };
        
        // Add the listener
        appointmentsQuery.addValueEventListener(appointmentsListener);
    }
    
    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        // Remove listener when activity is destroyed
        if (appointmentsListener != null && appointmentsQuery != null) {
            appointmentsQuery.removeEventListener(appointmentsListener);
        }
    }
} 

File: app/src/main/java/com/example/bloodbank/MyAppointmentsActivity.kt
package com.example.bloodbank

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.example.bloodbank.Adapter.AppointmentsAdapter
import com.example.bloodbank.Model.DonationAppointment
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

@AndroidEntryPoint
class MyAppointmentsActivity : AppCompatActivity() {
    private lateinit var appointmentsRecyclerView: RecyclerView
    private lateinit var adapter: AppointmentsAdapter
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var progressBar: ProgressBar 
    private lateinit var emptyView: TextView

    private val viewModel: MyAppointmentsViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_my_appointments)

        initializeViews()
        setupToolbar()
        setupRecyclerView()
        setupSwipeRefreshLayout()
        observeViewModel()

        // Trigger initial load
        viewModel.loadAppointments()
    }

    private fun initializeViews() {
        appointmentsRecyclerView = findViewById(R.id.appointmentsRecyclerView)
        swipeRefreshLayout = findViewById(R.id.swipeRefreshLayout)
        progressBar = findViewById(R.id.progressBar)
        emptyView = findViewById(R.id.emptyView)
    }

    private fun setupToolbar() {
        val toolbar: Toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            title = "My Appointments"
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowHomeEnabled(true)
        }
    }

    private fun setupRecyclerView() {
        // Initialize adapter with an empty list.
        adapter = AppointmentsAdapter(mutableListOf()) 
        appointmentsRecyclerView.layoutManager = LinearLayoutManager(this)
        appointmentsRecyclerView.adapter = adapter
    }

    private fun setupSwipeRefreshLayout() {
        swipeRefreshLayout.setOnRefreshListener {
            viewModel.loadAppointments()
        }
    }

    private fun observeViewModel() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                // Observe appointments
                launch {
                    viewModel.appointments.collectLatest { appointments ->
                        adapter.updateAppointments(appointments) 
                        emptyView.visibility = if (appointments.isEmpty()) View.VISIBLE else View.GONE
                    }
                }

                // Observe loading state
                launch {
                    viewModel.isLoading.collectLatest { isLoading ->
                        progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
                        swipeRefreshLayout.isRefreshing = isLoading
                    }
                }

                // Observe error events
                launch {
                    viewModel.error.collectLatest { errorMessage ->
                        Toast.makeText(this@MyAppointmentsActivity, errorMessage, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressedDispatcher.onBackPressed() 
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressedDispatcher.onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
